File: Data Structures & Algorithms/coin-change/submission-8.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        for(int i = 0;i<dp.length;i++){
            dp[i] = amount+1;
        }
        dp[0] = 0;
        for(int i = 1;i<=amount;i++){
            for(int c:coins){
                if(c<=i){
                    dp[i] = Math.min(dp[i],1+dp[i-c]);
                }
            }  
        }
        return dp[amount]<amount+1?dp[amount]:-1;


        // 0 = 0
        // 1 = 1 + dp[1-1]
        // 2 = 1 + dp[2-1]
        
    }
}

File: Data Structures & Algorithms/coin-change-ii/submission-8.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        coins.sort()
        N = len(coins)
        dp = {}
        def dfs(i,cur):
            if cur == amount:
                return 1
            if i == N or cur > amount:
                return 0
            if (i,cur) in dp:
                return dp[(i,cur)]
            
            dp[(i,cur)] = dfs(i+1,cur) + dfs(i,cur+coins[i])
            return dp[(i,cur)]
        return dfs(0,0)

        